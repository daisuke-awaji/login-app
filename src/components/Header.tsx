import React from 'react'
import { createStyles, makeStyles, Theme } from '@material-ui/core/styles'
import AppBar from '@material-ui/core/AppBar'
import Toolbar from '@material-ui/core/Toolbar'
import Typography from '@material-ui/core/Typography'
import IconButton from '@material-ui/core/IconButton'
import MenuIcon from '@material-ui/icons/Menu'
import { useDispatch } from 'react-redux'
import { unSetUser } from 'reducers/authenticate'
import ExitToAppIcon from '@material-ui/icons/ExitToApp'
import { Tooltip } from '@material-ui/core'
import AccountCircleIcon from '@material-ui/icons/AccountCircle'

const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    root: {
      flexGrow: 1,
    },
    menuButton: {
      marginRight: theme.spacing(2),
    },
    title: {
      flexGrow: 1,
    },
  }),
)

export default function Header() {
  const classes = useStyles()
  const dispatch = useDispatch()
  const handleClickLogout = () => dispatch(unSetUser())
  return (
    <div className={classes.root}>
      <AppBar position="static" color="primary">
        <Toolbar>
          <IconButton
            edge="start"
            className={classes.menuButton}
            color="inherit"
            aria-label="menu"
          >
            <MenuIcon />
          </IconButton>
          <Typography variant="h6" className={classes.title}>
            App
          </Typography>
          {/* Tooltip has Warnings in strict mode*/}
          {/* https://github.com/mui-org/material-ui/issues/13394 */}
          <Tooltip title="User Profile">
            <IconButton color="inherit">
              <AccountCircleIcon />
            </IconButton>
          </Tooltip>
          <Tooltip title="Logout">
            <IconButton color="inherit" onClick={handleClickLogout}>
              <ExitToAppIcon />
            </IconButton>
          </Tooltip>
        </Toolbar>
      </AppBar>
    </div>
  )
}
