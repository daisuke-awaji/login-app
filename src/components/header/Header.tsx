import React from 'react'
import { createStyles, makeStyles, Theme } from '@material-ui/core/styles'
import AppBar from '@material-ui/core/AppBar'
import Toolbar from '@material-ui/core/Toolbar'
import Typography from '@material-ui/core/Typography'
import IconButton from '@material-ui/core/IconButton'
import { useSelector } from 'react-redux'
import HomeIcon from '@material-ui/icons/Home'

import { Tooltip } from '@material-ui/core'
import AccountCircleIcon from '@material-ui/icons/AccountCircle'
import { useHistory } from 'react-router-dom'
import { ServiceLogo } from '../ServiceLogo'
import LogoutButton from '../logout/LogoutButton'

const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    root: {
      flexGrow: 1,
    },
    menuButton: {
      marginRight: theme.spacing(2),
    },
    title: {
      flexGrow: 1,
    },
  }),
)

/**
 * Tooltip has Warnings in strict mode
 * https://github.com/mui-org/material-ui/issues/13394
 */
const HomeButton = () => {
  const history = useHistory()
  const handleClickHome = () => history.push('/')
  return (
    <Tooltip title="Home">
      <IconButton color="inherit" onClick={handleClickHome}>
        <HomeIcon />
      </IconButton>
    </Tooltip>
  )
}

const AppIconButton = (props: any) => {
  const { className } = props
  const history = useHistory()
  const handleClickHome = () => history.push('/')
  return (
    <IconButton
      edge="start"
      className={className}
      color="inherit"
      aria-label="home"
      onClick={handleClickHome}
    >
      <ServiceLogo />
    </IconButton>
  )
}

const UserProfileButton = () => {
  const user = useSelector((state: any) => state.authenticatedUser)
  const history = useHistory()
  const handleClick = () => history.push(`/users/${user.id}`, { update: true })
  const title = `User Profile: ${user?.name}`
  return (
    <Tooltip title={title}>
      <IconButton color="inherit" onClick={handleClick}>
        <AccountCircleIcon />
      </IconButton>
    </Tooltip>
  )
}

export default function Header() {
  const classes = useStyles()
  return (
    <div className={classes.root}>
      <AppBar
        position="static"
        color="primary"
        style={{ background: 'transparent', boxShadow: 'none' }}
      >
        <Toolbar>
          <Typography variant="h6" className={classes.title}>
            <AppIconButton className={classes.menuButton} />
          </Typography>
          <HomeButton />
          <UserProfileButton />
          <LogoutButton />
        </Toolbar>
      </AppBar>
    </div>
  )
}
