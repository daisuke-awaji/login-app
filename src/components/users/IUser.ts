export interface IUser {
  id: string
  name: string
  email: string
  gender?: Gender
  type: UserType
  description?: string
  img?: string
}

export type Gender = 'mail' | 'femail'
export type UserType = 'admin' | 'default'
