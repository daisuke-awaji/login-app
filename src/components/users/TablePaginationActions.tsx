import React from 'react'
import { Theme, createStyles, makeStyles } from '@material-ui/core/styles'
import IconButton from '@material-ui/core/IconButton'
import FirstPageIcon from '@material-ui/icons/FirstPage'
import KeyboardArrowLeft from '@material-ui/icons/KeyboardArrowLeft'
import KeyboardArrowRight from '@material-ui/icons/KeyboardArrowRight'
import LastPageIcon from '@material-ui/icons/LastPage'

const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    root: {
      flexShrink: 0,
      marginLeft: theme.spacing(2.5),
    },
  }),
)
interface TablePaginationActionsProps {
  count: number
  page: number
  rowsPerPage: number
  onChangePage: (
    event: React.MouseEvent<HTMLButtonElement>,
    newPage: number,
  ) => void
}

/**
 *
 * 表の下部に配置する |< < > >| というボタンコンポーネント
 * @param props
 */
export function TablePaginationActions(props: TablePaginationActionsProps) {
  const classes = useStyles()
  const { count, page, rowsPerPage, onChangePage } = props
  return (
    <div className={classes.root}>
      <IconButton
        onClick={(event) => onChangePage(event, 0)}
        disabled={page === 0}
        aria-label="first page |<"
      >
        <FirstPageIcon />
      </IconButton>
      <IconButton
        onClick={(event) => onChangePage(event, page - 1)}
        disabled={page === 0}
        aria-label="previous page <"
      >
        <KeyboardArrowLeft />
      </IconButton>
      <IconButton
        onClick={(event) => onChangePage(event, page + 1)}
        disabled={page >= Math.ceil(count / rowsPerPage) - 1}
        aria-label="next page >"
      >
        <KeyboardArrowRight />
      </IconButton>
      <IconButton
        onClick={(event) => {
          const lastPage = Math.max(0, Math.ceil(count / rowsPerPage) - 1)
          onChangePage(event, lastPage)
        }}
        disabled={page >= Math.ceil(count / rowsPerPage) - 1}
        aria-label="last page >|"
      >
        <LastPageIcon />
      </IconButton>
    </div>
  )
}
