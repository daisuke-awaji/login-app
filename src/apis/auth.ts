import { IUser } from 'components/users/IUser'

const sleep = (s: number) => new Promise((resolve) => setTimeout(resolve, s))

export async function login({ email, password }: any): Promise<IUser> {
  await sleep(500)
  return new Promise((resolve, reject) => {
    if (email === 'usera@email.com' && password === 'a') {
      resolve({
        id: 123,
        name: 'John Scott',
        email: email,
        type: 'admin',
      })
    } else if (email === 'userb@email.com' && password === 'b') {
      resolve({
        id: 456,
        name: 'Marty',
        email: email,
        type: 'default',
      })
    } else {
      reject(new Error('email or password is not valid'))
    }
  })
}

export async function currentUser(sessionId: string | null): Promise<IUser> {
  await sleep(500)

  return new Promise((resolve, reject) => {
    if (sessionId === '1') {
      resolve({
        id: 123,
        name: 'John Scott',
        email: 'example@email.com',
        type: 'admin',
      })
    } else {
      reject(new Error('unauthorized'))
    }
  })
}
