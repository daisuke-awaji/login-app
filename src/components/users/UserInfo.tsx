import React, { useEffect, useState } from 'react'
import { useParams } from 'react-router-dom'
import { fetchUsers } from 'apis/users'
import { IUser } from './IUser'

export const UserInfo = () => {
  // TODO: 本当はユーザ１件取得のAPIを呼ぶ
  const [users, setUsers] = useState<IUser[]>([])
  const fetchData = async () => {
    const users = await fetchUsers()
    setUsers(users)
  }
  useEffect(() => {
    fetchData()
  }, [])

  const { id } = useParams()
  const one = users.find((user) => user.id === Number(id))
  return (
    <div>
      <h1>{one?.name}</h1>
      <h3>Email: {one?.email}</h3>
    </div>
  )
}
