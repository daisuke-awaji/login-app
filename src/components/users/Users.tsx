import React, { useEffect, useState } from 'react'
import { Link } from 'react-router-dom'
import { fetchUsers } from 'apis/users'

import Table from '@material-ui/core/Table'
import TableBody from '@material-ui/core/TableBody'
import TableCell from '@material-ui/core/TableCell'
import TableContainer from '@material-ui/core/TableContainer'
import TableHead from '@material-ui/core/TableHead'
import TableRow from '@material-ui/core/TableRow'
import Paper from '@material-ui/core/Paper'
import Typography from '@material-ui/core/Typography'
import { IUser } from './IUser'

export const Users = () => {
  const [users, setUsers] = useState<IUser[]>([])

  const fetchData = async () => {
    const users = await fetchUsers()
    setUsers(users)
  }
  useEffect(() => {
    fetchData()
  }, [])
  return (
    <>
      <Typography variant="h3">Users</Typography>
      <TableContainer component={Paper}>
        <Table style={{ tableLayout: 'auto' }} aria-label="user table">
          <TableHead>
            <TableRow>
              <TableCell style={{ width: '10%' }} align="right">
                Id
              </TableCell>
              <TableCell style={{ width: '30%' }} align="right">
                Name
              </TableCell>
              <TableCell style={{ width: '40%' }} align="right">
                Email
              </TableCell>
              <TableCell style={{ width: '10%' }} align="right">
                Type
              </TableCell>
            </TableRow>
          </TableHead>
          <TableBody>
            {users.map((user) => (
              <TableRow key={user.name}>
                <TableCell style={{ width: '10%' }} align="right">
                  {user.id}
                </TableCell>
                <TableCell style={{ width: '30%' }} align="right">
                  <Link to={'/users/' + user.id}>{user.name}</Link>
                </TableCell>
                <TableCell
                  style={{
                    width: '40%',
                    maxWidth: '100px',
                    // textOverflow: 'ellipsis', // はみ出たテキストを切り取って ... を表示する
                    whiteSpace: 'nowrap',
                    overflow: 'scroll',
                  }}
                  align="right"
                >
                  {user.email}
                </TableCell>
                <TableCell style={{ width: '10%' }} align="right">
                  {user.type}
                </TableCell>
              </TableRow>
            ))}
          </TableBody>
        </Table>
      </TableContainer>
    </>
  )
}
