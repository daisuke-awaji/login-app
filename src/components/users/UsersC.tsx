import React from 'react'
import { Link } from 'react-router-dom'
import { fetchUsers } from 'apis/users'
import { IUser } from './IUser'

import Table from '@material-ui/core/Table'
import TableBody from '@material-ui/core/TableBody'
import TableCell from '@material-ui/core/TableCell'
import TableContainer from '@material-ui/core/TableContainer'
import TableHead from '@material-ui/core/TableHead'
import TableRow from '@material-ui/core/TableRow'
import Paper from '@material-ui/core/Paper'
import Typography from '@material-ui/core/Typography'

interface IProps {
  classes?: any
}
interface IState {
  users: IUser[]
}

export class Users extends React.Component<IProps, IState> {
  constructor(props: any) {
    super(props)
    this.state = { users: [] }
  }
  async componentDidMount() {
    const users = await fetchUsers()
    this.setState({ users: users })
  }
  render() {
    return (
      <>
        <Typography variant="h3">Users</Typography>
        <TableContainer component={Paper}>
          <Table style={{ tableLayout: 'auto' }} aria-label="user table">
            <TableHead>
              <TableRow>
                <TableCell style={{ width: '10%' }} align="right">
                  Id
                </TableCell>
                <TableCell style={{ width: '30%' }} align="right">
                  Name
                </TableCell>
                <TableCell style={{ width: '40%' }} align="right">
                  Email
                </TableCell>
                <TableCell style={{ width: '10%' }} align="right">
                  Type
                </TableCell>
              </TableRow>
            </TableHead>
            <TableBody>
              {this.state.users.map((user) => (
                <TableRow key={user.name}>
                  <TableCell style={{ width: '10%' }} align="right">
                    {user.id}
                  </TableCell>
                  <TableCell style={{ width: '30%' }} align="right">
                    <Link to={'/users/' + user.id}>{user.name}</Link>
                  </TableCell>
                  <TableCell
                    style={{
                      width: '40%',
                      maxWidth: '100px',
                      // textOverflow: 'ellipsis', // はみ出たテキストを切り取って ... を表示する
                      whiteSpace: 'nowrap',
                      overflow: 'scroll',
                    }}
                    align="right"
                  >
                    {user.email}
                  </TableCell>
                  <TableCell style={{ width: '10%' }} align="right">
                    {user.type}
                  </TableCell>
                </TableRow>
              ))}
            </TableBody>
          </Table>
        </TableContainer>
      </>
    )
  }
}
