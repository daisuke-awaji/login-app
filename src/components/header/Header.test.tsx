import React from 'react'
import { shallow, mount } from 'enzyme'
import { Provider } from 'react-redux'
import { Router } from 'react-router-dom'
import { combineReducers, createStore } from 'redux'
import auth, { loginSucceedAction } from 'reducers/authenticate'
import Header from './Header'
import renderer from 'react-test-renderer'
import { IUser } from '../users/IUser'

describe('学習用テストコード', () => {
  test('Enzyme shallow', () => {
    const element = <a href="http://www.facebook.com">Facebook</a>
    const wrapper = shallow(element)
    expect(wrapper.find('a').prop('href')).toBe('http://www.facebook.com')
  })
})

const setup = () => {
  // Routerコンポーネントに依存する hisotry は mock 化して注入する
  const mockHistory: any = { push: jest.fn(), location: {}, listen: jest.fn() }
  // Redux は必要な store だけ mock を作成して注入する
  const reducer = combineReducers({ authenticatedUser: auth })
  const mockStore = createStore(reducer)
  const component = (
    <Provider store={mockStore}>
      <Router history={mockHistory}>
        <Header />
      </Router>
    </Provider>
  )
  const getWrapper = () => mount(component)

  const user: IUser = {
    id: '1',
    name: 'John Scott',
    email: 'john@email.com',
    type: 'admin',
  }
  mockStore.dispatch(loginSucceedAction(user))
  return { wrapper: getWrapper(), component }
}
/**
 * Enzymeのshallowを使っても良いが
 * Reduxのstoreを剥がす実装コストを考慮すると
 * mountを使用したFULLレンダリングの方が効率的で可読性も良いと考えた
 */
describe('Headerコンポーネントの FULLレンダリングテスト', () => {
  describe('Homeボタン', () => {
    it('Homeボタン押下でHome画面に遷移される', () => {
      const { wrapper } = setup()
      wrapper.find('button').at(0).simulate('click')
      expect(window.location.pathname).toEqual('/')
    })
  })
})

/**
 * Headerのような大きなサイズのコンポーネントに対してはスナップショットテストで十分
 */
describe('スナップショット', () => {
  it('レンダリングができること', () => {
    const { component } = setup()
    const header = renderer.create(component)
    let tree = header.toJSON()
    expect(tree).toMatchSnapshot()
  })
})
