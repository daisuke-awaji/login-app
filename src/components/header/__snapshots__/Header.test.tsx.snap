// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`スナップショット レンダリングができること 1`] = `
<div
  className="makeStyles-root-1"
>
  <header
    className="MuiPaper-root MuiAppBar-root MuiAppBar-positionStatic MuiAppBar-colorPrimary MuiPaper-elevation4"
    style={
      Object {
        "background": "transparent",
        "boxShadow": "none",
      }
    }
  >
    <div
      className="MuiToolbar-root MuiToolbar-regular MuiToolbar-gutters"
    >
      <h6
        className="MuiTypography-root makeStyles-title-3 MuiTypography-h6"
      >
        <button
          aria-label="home"
          className="MuiButtonBase-root MuiIconButton-root makeStyles-menuButton-2 MuiIconButton-colorInherit MuiIconButton-edgeStart"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="button"
        >
          <span
            className="MuiIconButton-label"
          >
            <img
              alt="logo"
              src="logo.png"
              style={
                Object {
                  "height": "48px",
                }
              }
            />
          </span>
          <span
            className="MuiTouchRipple-root"
          />
        </button>
      </h6>
      <p
        className="MuiTypography-root MuiTypography-body1"
      >
        John Scott
      </p>
      <button
        aria-describedby={null}
        className="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseOver={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        title="Home"
        type="button"
      >
        <span
          className="MuiIconButton-label"
        >
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M10 20v-6h4v6h5v-8h3L12 3 2 12h3v8z"
            />
          </svg>
        </span>
        <span
          className="MuiTouchRipple-root"
        />
      </button>
      <button
        aria-describedby={null}
        className="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit"
        disabled={false}
        onBlur={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseOver={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        title="User Profile"
        type="button"
      >
        <span
          className="MuiIconButton-label"
        >
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 3c1.66 0 3 1.34 3 3s-1.34 3-3 3-3-1.34-3-3 1.34-3 3-3zm0 14.2c-2.5 0-4.71-1.28-6-3.22.03-1.99 4-3.08 6-3.08 1.99 0 5.97 1.09 6 3.08-1.29 1.94-3.5 3.22-6 3.22z"
            />
          </svg>
        </span>
        <span
          className="MuiTouchRipple-root"
        />
      </button>
      <button
        aria-describedby={null}
        className="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseOver={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        title="Logout"
        type="button"
      >
        <span
          className="MuiIconButton-label"
        >
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M10.09 15.59L11.5 17l5-5-5-5-1.41 1.41L12.67 11H3v2h9.67l-2.58 2.59zM19 3H5c-1.11 0-2 .9-2 2v4h2V5h14v14H5v-4H3v4c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
            />
          </svg>
        </span>
        <span
          className="MuiTouchRipple-root"
        />
      </button>
    </div>
  </header>
</div>
`;
